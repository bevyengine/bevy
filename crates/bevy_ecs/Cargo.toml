[package]
name = "bevy_ecs"
version = "0.4.0"
edition = "2018"
authors = [
    "Bevy Contributors <bevyengine@gmail.com>",
    "Carter Anderson <mcanders1@gmail.com>",
]
description = "Bevy Engine's entity component system"
homepage = "https://bevyengine.org"
repository = "https://github.com/bevyengine/bevy"
license = "Apache-2.0"
keywords = ["ecs", "game", "bevy"]
categories = ["game-engines", "data-structures"]

[features]
trace = []

[dependencies]
bevy_tasks = { path = "../bevy_tasks", version = "0.4.0" }
bevy_utils = { path = "../bevy_utils", version = "0.4.0" }
bevy_ecs_macros = { path = "macros", version = "0.4.0" }
fxhash = "0.2"
rand = "0.8.0"
serde = "1.0"
thiserror = "1.0"
fixedbitset = "0.3.1"
bitflags = "1.2.1"
downcast-rs = "1.2.0"
parking_lot = "0.11.0"
lazy_static = { version = "1.4.0" }

[dev-dependencies]
bencher = "0.1.5"

[[bench]]
name = "bench"
harness = false
required-features = ["macros"]
