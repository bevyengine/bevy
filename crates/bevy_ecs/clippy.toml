disallowed-methods = [
  { path = "UniqueEntityVec::from_iter", reason = "Use UniqueEntityVec::from_entity_set_iter if possible, it skips validation" },
  { path = "EntitySetIterator::collect::<UniqueEntityVec>", reason = "Use EntitySetIterator::collect_set if possible, it skips validation" },
]
