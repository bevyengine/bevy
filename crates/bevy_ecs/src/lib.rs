pub use bevy_hecs::{Query as HecsQuery, *};
mod resource;
mod schedule;
mod system;
mod world;

pub use resource::*;
pub use schedule::*;
pub use system::{Query, *};
pub use world::*;

pub mod prelude {
    pub use crate::{
        resource::{ChangedRes, FromResources, Local, OrRes, Res, ResMut, Resource, Resources},
        system::{
            Commands, IntoForEachSystem, IntoQuerySystem, IntoThreadLocalSystem, Query, System,
        },
        world::WorldBuilderSource,
        Added, Bundle, Changed, Component, Entity, Mut, Mutated, Or, QuerySet, Ref, RefMut, With,
        Without, World,
    };
}
