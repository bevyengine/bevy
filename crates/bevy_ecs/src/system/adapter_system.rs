use std::borrow::Cow;

use super::{ReadOnlySystem, System};
use crate::{schedule::InternedSystemSet, world::unsafe_world_cell::UnsafeWorldCell};

/// Customizes the behavior of an [`AdapterSystem`]
///
/// # Examples
///
/// ```
/// # use bevy_ecs::prelude::*;
/// use bevy_ecs::system::{Adapt, AdapterSystem};
///
/// // A system adapter that inverts the result of a system.
/// // NOTE: Instead of manually implementing this, you can just use `bevy_ecs::schedule::common_conditions::not`.
/// pub type NotSystem<S> = AdapterSystem<NotMarker, S>;
///
/// // This struct is used to customize the behavior of our adapter.
/// pub struct NotMarker;
///
/// impl<S> Adapt<S> for NotMarker
/// where
///     S: System,
///     S::Out: std::ops::Not,
/// {
///     type In = S::In;
///     type Out = <S::Out as std::ops::Not>::Output;
///
///     fn adapt(
///         &mut self,
///         input: Self::In,
///         run_system: impl FnOnce(S::In) -> S::Out,
///     ) -> Self::Out {
///         !run_system(input)
///     }
/// }
/// # let mut world = World::new();
/// # let mut system = NotSystem::new(NotMarker, IntoSystem::into_system(|| false), "".into());
/// # system.initialize(&mut world);
/// # assert!(system.run((), &mut world));
/// ```
pub trait Adapt<S: System>: Send + Sync + 'static {
    /// The [input](System::In) type for an [`AdapterSystem`].
    type In;
    /// The [output](System::Out) type for an [`AdapterSystem`].
    type Out;

    /// When used in an [`AdapterSystem`], this function customizes how the system
    /// is run and how its inputs/outputs are adapted.
    fn adapt(&mut self, input: Self::In, run_system: impl FnOnce(S::In) -> S::Out) -> Self::Out;
}

/// A [`System`] that takes the output of `S` and transforms it by applying `Func` to it.
#[derive(Clone)]
pub struct AdapterSystem<Func, S> {
    func: Func,
    system: S,
    name: Cow<'static, str>,
}

impl<Func, S> AdapterSystem<Func, S>
where
    Func: Adapt<S>,
    S: System,
{
    /// Creates a new [`System`] that uses `func` to adapt `system`, via the [`Adapt`] trait.
    pub const fn new(func: Func, system: S, name: Cow<'static, str>) -> Self {
        Self { func, system, name }
    }
}

impl<Func, S> System for AdapterSystem<Func, S>
where
    Func: Adapt<S>,
    S: System,
{
    type In = Func::In;
    type Out = Func::Out;

    fn name(&self) -> Cow<'static, str> {
        self.name.clone()
    }

    fn component_access(&self) -> &crate::query::Access<crate::component::ComponentId> {
        self.system.component_access()
    }

    #[inline]
    fn archetype_component_access(
        &self,
    ) -> &crate::query::Access<crate::archetype::ArchetypeComponentId> {
        self.system.archetype_component_access()
    }

    fn is_send(&self) -> bool {
        self.system.is_send()
    }

    fn is_exclusive(&self) -> bool {
        self.system.is_exclusive()
    }

    fn has_deferred(&self) -> bool {
        self.system.has_deferred()
    }

    #[inline]
    unsafe fn run_unsafe(&mut self, input: Self::In, world: UnsafeWorldCell) -> Self::Out {
        // SAFETY: `system.run_unsafe` has the same invariants as `self.run_unsafe`.
        self.func.adapt(input, |input| unsafe {
            self.system.run_unsafe(input, world)
        })
    }

    #[inline]
    fn run(&mut self, input: Self::In, world: &mut crate::prelude::World) -> Self::Out {
        self.func
            .adapt(input, |input| self.system.run(input, world))
    }

    #[inline]
    fn apply_deferred(&mut self, world: &mut crate::prelude::World) {
        self.system.apply_deferred(world);
    }

    fn initialize(&mut self, world: &mut crate::prelude::World) {
        self.system.initialize(world);
    }

    #[inline]
    fn update_archetype_component_access(&mut self, world: UnsafeWorldCell) {
        self.system.update_archetype_component_access(world);
    }

    fn check_change_tick(&mut self, change_tick: crate::component::Tick) {
        self.system.check_change_tick(change_tick);
    }

    fn default_system_sets(&self) -> Vec<InternedSystemSet> {
        self.system.default_system_sets()
    }

    fn get_last_run(&self) -> crate::component::Tick {
        self.system.get_last_run()
    }

    fn set_last_run(&mut self, last_run: crate::component::Tick) {
        self.system.set_last_run(last_run);
    }
}

// SAFETY: The inner system is read-only.
unsafe impl<Func, S> ReadOnlySystem for AdapterSystem<Func, S>
where
    Func: Adapt<S>,
    S: ReadOnlySystem,
{
}

impl<F, S, Out> Adapt<S> for F
where
    S: System,
    F: Send + Sync + 'static + FnMut(S::Out) -> Out,
{
    type In = S::In;
    type Out = Out;

    fn adapt(&mut self, input: S::In, run_system: impl FnOnce(S::In) -> S::Out) -> Out {
        self(run_system(input))
    }
}
