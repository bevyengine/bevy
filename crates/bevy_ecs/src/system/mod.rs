mod commands;
mod into_system;
mod into_thread_local;
mod query;
#[allow(clippy::module_inception)]
mod system;
mod system_chaining;
mod system_param;

pub use commands::*;
pub use into_system::*;
pub use into_thread_local::*;
pub use query::*;
pub use system::*;
pub use system_chaining::*;
pub use system_param::*;
