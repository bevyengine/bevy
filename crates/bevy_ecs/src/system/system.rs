use crate::resource::Resources;
use bevy_hecs::{ArchetypeComponent, TypeAccess, World};
use std::{any::TypeId, borrow::Cow};

/// Determines the strategy used to run the `run_thread_local` function in a [System]
#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub enum ThreadLocalExecution {
    Immediate,
    NextFlush,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
pub struct SystemId(pub usize);

impl SystemId {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        SystemId(rand::random::<usize>())
    }
}

/// An ECS system that can be added to a [Schedule](crate::Schedule)
pub trait System: Send + Sync {
    fn name(&self) -> Cow<'static, str>;
    fn id(&self) -> SystemId;
    fn update(&mut self, world: &World);
    fn archetype_component_access(&self) -> &TypeAccess<ArchetypeComponent>;
    fn resource_access(&self) -> &TypeAccess<TypeId>;
    fn thread_local_execution(&self) -> ThreadLocalExecution;
    fn run(&mut self, world: &World, resources: &Resources);
    fn run_thread_local(&mut self, world: &mut World, resources: &mut Resources);
    fn initialize(&mut self, _world: &mut World, _resources: &mut Resources) {}
}
