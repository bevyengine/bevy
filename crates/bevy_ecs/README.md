# Bevy ECS

[![Crates.io](https://img.shields.io/crates/v/bevy_ecs.svg)](https://crates.io/crates/bevy_ecs)
[![license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/bevyengine/bevy/blob/HEAD/LICENSE)
[![Discord](https://img.shields.io/discord/691052431525675048.svg?label=&logo=discord&logoColor=ffffff&color=7389D8&labelColor=6A7EC2)](https://discord.gg/bevy)

## What is Bevy ECS?

Bevy ECS is an Entity Component System custom-built for the [Bevy][bevy] game engine. It aims to be simple to use, ergonomic, fast, massively parallel, opinionated, and featureful. It was created specifically for Bevy's needs, but it can easily be used as a standalone crate in other projects.

## ECS

All app logic in Bevy uses the Entity Component System paradigm, which is often shortened to ECS. ECS is a software pattern that involves breaking your program up into Entities, Components, and Systems. Entities are unique "things" that are assigned groups of Components, which are then processed using Systems.

For example, one entity might have a `Position` and `Velocity` component, whereas another entity might have a `Position` and `UI` component. You might have a movement system that runs on all entities with a Position and Velocity component.

The ECS pattern encourages clean, decoupled designs by forcing you to break up your app data and logic into its core components. It also helps make your code faster by optimizing memory access patterns and making parallelism easier.

## Concepts

Bevy ECS is Bevy's implementation of the ECS pattern. Unlike other Rust ECS implementations, which often require complex lifetimes, traits, builder patterns, or macros, Bevy ECS uses normal Rust data types for all of these concepts:

### Components

Components are normal Rust structs. They are data stored in a `World` and specific instances of Components correlate to Entities.

```rust
use bevy_ecs::prelude::*;

#[derive(Component)]
struct Position { x: f32, y: f32 }
```

### Worlds

Entities, Components, and Resources are stored in a `World`. Worlds, much like Rust std collections like HashSet and Vec, expose operations to insert, read, write, and remove the data they store.

```rust
use bevy_ecs::world::World;

let world = World::default();
```

### Entities

Entities are unique identifiers that correlate to zero or more Components.

```rust
use bevy_ecs::prelude::*;

#[derive(Component)]
struct Position { x: f32, y: f32 }
#[derive(Component)]
struct Velocity { x: f32, y: f32 }

let mut world = World::new();

let entity = world.spawn()
    .insert(Position { x: 0.0, y: 0.0 })
    .insert(Velocity { x: 1.0, y: 0.0 })
    .id();

let entity_ref = world.entity(entity);
let position = entity_ref.get::<Position>().unwrap();
let velocity = entity_ref.get::<Velocity>().unwrap();
```

### Systems

Systems are normal Rust functions. Thanks to the Rust type system, Bevy ECS can use function parameter types to determine what data needs to be sent to the system. It also uses this "data access" information to determine what Systems can run in parallel with each other.

```rust
use bevy_ecs::prelude::*;

#[derive(Component)]
struct Position { x: f32, y: f32 }

fn print_position(query: Query<(Entity, &Position)>) {
    for (entity, position) in query.iter() {
        println!("Entity {:?} is at position: x {}, y {}", entity, position.x, position.y);
    }
}
```

### Resources

Apps often require unique resources, such as asset collections, renderers, audio servers, time, etc. Bevy ECS makes this pattern a first class citizen. `Resource` is a special kind of component that does not belong to any entity. Instead, it is identified uniquely by its type:

```rust
use bevy_ecs::prelude::*;

#[derive(Default)]
struct Time {
    seconds: f32,
}

let mut world = World::new();

world.insert_resource(Time::default());

let time = world.get_resource::<Time>().unwrap();

// You can also access resources from Systems
fn print_time(time: Res<Time>) {
    println!("{}", time.seconds);
}
```

The [`resources.rs`](examples/resources.rs) example illustrates how to read and write a Counter resource from Systems.

### Schedules

Schedules consist of zero or more Stages, which run a set of Systems according to some execution strategy. Bevy ECS provides a few built in Stage implementations (ex: parallel, serial), but you can also implement your own! Schedules run Stages one-by-one in an order defined by the user.

The built in "parallel stage" considers dependencies between systems and (by default) run as many of them in parallel as possible. This maximizes performance, while keeping the system execution safe. You can also define explicit dependencies between systems.

## Using Bevy ECS

Bevy ECS should feel very natural for those familiar with Rust syntax:

```rust
use bevy_ecs::prelude::*;

#[derive(Component)]
struct Position { x: f32, y: f32 }
#[derive(Component)]
struct Velocity { x: f32, y: f32 }

// This system moves each entity with a Position and Velocity component
fn movement(mut query: Query<(&mut Position, &Velocity)>) {
    for (mut position, velocity) in query.iter_mut() {
        position.x += velocity.x;
        position.y += velocity.y;
    }
}

fn main() {
    // Create a new empty World to hold our Entities and Components
    let mut world = World::new();

    // Spawn an entity with Position and Velocity components
    world.spawn()
        .insert(Position { x: 0.0, y: 0.0 })
        .insert(Velocity { x: 1.0, y: 0.0 });

    // Create a new Schedule, which defines an execution strategy for Systems
    let mut schedule = Schedule::default();

    // Add a Stage to our schedule. Each Stage in a schedule runs all of its systems
    // before moving on to the next Stage
    schedule.add_stage("update", SystemStage::parallel()
        .with_system(movement)
    );

    // Run the schedule once. If your app has a "loop", you would run this once per loop
    schedule.run(&mut world);
}
```

## Features

### Query Filters

```rust
use bevy_ecs::prelude::*;

#[derive(Component)]
struct Position { x: f32, y: f32 }
#[derive(Component)]
struct Player;
#[derive(Component)]
struct Alive;

// Gets the Position component of all Entities with Player component and without the Alive
// component. 
fn system(query: Query<&Position, (With<Player>, Without<Alive>)>) {
    for position in query.iter() {
    }
}
```

### Change Detection

Bevy ECS tracks _all_ changes to Components and Resources.

Queries can filter for changed Components:

```rust
use bevy_ecs::prelude::*;

#[derive(Component)]
struct Position { x: f32, y: f32 }
#[derive(Component)]
struct Velocity { x: f32, y: f32 }

// Gets the Position component of all Entities whose Velocity has changed since the last run of the System
fn system_changed(query: Query<&Position, Changed<Velocity>>) {
    for position in query.iter() {
    }
}

// Gets the i32 component of all Entities that had a f32 component added since the last run of the System
fn system_added(query: Query<&Position, Added<Velocity>>) {
    for position in query.iter() {
    }
}
```

Resources also expose change state:

```rust
use bevy_ecs::prelude::*;

struct Time(f32);

// Prints "time changed!" if the Time resource has changed since the last run of the System
fn system(time: Res<Time>) {
    if time.is_changed() {
        println!("time changed!");
    }
}
```

The [`change_detection.rs`](examples/change_detection.rs) example shows how to query only for updated entities and react on changes in resources.

### Component Storage

Bevy ECS supports multiple component storage types.

Components can be stored in:

* **Tables**: Fast and cache friendly iteration, but slower adding and removing of components. This is the default storage type.
* **Sparse Sets**: Fast adding and removing of components, but slower iteration.

Component storage types are configurable, and they default to table storage if the storage is not manually defined.

```rust
use bevy_ecs::prelude::*;

#[derive(Component)]
struct TableStoredComponent;

#[derive(Component)]
#[component(storage = "SparseSet")]
struct SparseStoredComponent;
```

### Component Bundles

Define sets of Components that should be added together.

```rust
use bevy_ecs::prelude::*;

#[derive(Default, Component)]
struct Player;
#[derive(Default, Component)]
struct Position { x: f32, y: f32 }
#[derive(Default, Component)]
struct Velocity { x: f32, y: f32 }

#[derive(Bundle, Default)]
struct PlayerBundle {
    player: Player,
    position: Position,
    velocity: Velocity,
}

let mut world = World::new();

// Spawn a new entity and insert the default PlayerBundle
world.spawn().insert_bundle(PlayerBundle::default());

// Bundles play well with Rust's struct update syntax
world.spawn().insert_bundle(PlayerBundle {
    position: Position { x: 1.0, y: 1.0 },
    ..Default::default()
});
```

### Events

Events offer a communication channel between one or more systems. Events can be sent using the system parameter `EventWriter` and received with `EventReader`.

```rust
use bevy_ecs::prelude::*;

struct MyEvent {
    message: String,
}

fn writer(mut writer: EventWriter<MyEvent>) {
    writer.send(MyEvent {
        message: "hello!".to_string(),
    });
}

fn reader(mut reader: EventReader<MyEvent>) {
    for event in reader.iter() {
    }
}
```

A minimal set up using events can be seen in [`events.rs`](examples/events.rs).

[bevy]: https://bevyengine.org/
