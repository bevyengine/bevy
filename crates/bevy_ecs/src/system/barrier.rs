use bevy_utils::all_tuples;

use crate::component::Tick;
use crate::prelude::World;
use crate::schedule::{InternedSystemSet, SystemSet};
use crate::system::{SystemMeta, SystemParam};
use crate::world::unsafe_world_cell::UnsafeWorldCell;
use std::marker::PhantomData;

/// List of barrier dependencies for the system.
pub trait BarrierList {
    /// System sets before this system.
    fn before_list() -> Vec<InternedSystemSet>;
    /// System sets after this system.
    fn after_list() -> Vec<InternedSystemSet>;
}

/// System param to mark current system to run before a given system set.
pub struct Before<S: SystemSet + Default> {
    marker: PhantomData<S>,
}

/// System param to mark current system to run after a given system set.
pub struct After<S: SystemSet + Default> {
    marker: PhantomData<S>,
}

unsafe impl<S: SystemSet + Default> SystemParam for Before<S> {
    type State = ();
    type Item<'world, 'state> = Self;
    type BarrierList = BeforeBarrierList<S>;

    fn init_state(_world: &mut World, _system_meta: &mut SystemMeta) -> Self::State {
        ()
    }

    unsafe fn get_param<'world, 'state>(
        _state: &'state mut Self::State,
        _system_meta: &SystemMeta,
        _world: UnsafeWorldCell<'world>,
        _change_tick: Tick,
    ) -> Self::Item<'world, 'state> {
        Before {
            marker: PhantomData,
        }
    }
}

unsafe impl<S: SystemSet + Default> SystemParam for After<S> {
    type State = ();
    type Item<'world, 'state> = Self;

    type BarrierList = AfterBarrierList<S>;

    fn init_state(_world: &mut World, _system_meta: &mut SystemMeta) -> Self::State {
        ()
    }

    unsafe fn get_param<'world, 'state>(
        _state: &'state mut Self::State,
        _system_meta: &SystemMeta,
        _world: UnsafeWorldCell<'world>,
        _change_tick: Tick,
    ) -> Self::Item<'world, 'state> {
        After {
            marker: PhantomData,
        }
    }
}

#[doc(hidden)]
pub struct BeforeBarrierList<S: SystemSet + Default>(PhantomData<S>);
#[doc(hidden)]
pub struct AfterBarrierList<S: SystemSet + Default>(PhantomData<S>);

impl<S: SystemSet + Default> BarrierList for BeforeBarrierList<S> {
    fn before_list() -> Vec<InternedSystemSet> {
        vec![S::default().intern()]
    }

    fn after_list() -> Vec<InternedSystemSet> {
        Vec::new()
    }
}

impl<S: SystemSet + Default> BarrierList for AfterBarrierList<S> {
    fn before_list() -> Vec<InternedSystemSet> {
        Vec::new()
    }

    fn after_list() -> Vec<InternedSystemSet> {
        vec![S::default().intern()]
    }
}

macro_rules! barrier_list_for_tuple {
    ($($barrier: ident),*) => {
        impl<$($barrier),*> BarrierList for ($($barrier,)*)
        where
            $($barrier: BarrierList),*
        {
            fn before_list() -> Vec<InternedSystemSet> {
                let mut list = Vec::new();
                let _ignore_empty = &mut list;
                $(list.extend($barrier::before_list());)*
                list
            }

            fn after_list() -> Vec<InternedSystemSet> {
                let mut list = Vec::new();
                let _ignore_empty = &mut list;
                $(list.extend($barrier::after_list());)*
                list
            }
        }
    }
}

all_tuples!(barrier_list_for_tuple, 0, 16, P);
