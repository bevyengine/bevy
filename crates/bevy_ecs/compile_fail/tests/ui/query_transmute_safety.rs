use bevy_ecs::prelude::*;
use bevy_ecs::system::SystemState;

#[derive(Component, Eq, PartialEq, Debug)]
struct Foo(u32);

#[derive(Component)]
struct Bar;

fn main() {
    let mut world = World::default();
    world.spawn(Foo(10));

    let mut system_state = SystemState::<Query<(&mut Foo, &Bar)>>::new(&mut world);
    let mut query = system_state.get_mut(&mut world);

    {
        let mut lens_a = query.transmute_lens::<&mut Foo>();
        let mut lens_b = query.transmute_lens::<&mut Foo>();
        //~^ E0499

        let mut query_a = lens_a.query();
        let mut query_b = lens_b.query();

        let a = query_a.single_mut();
        let b = query_b.single_mut(); // oops 2 mutable references to same Foo
        assert_eq!(*a, *b);
    }

    {
        let mut lens = query.transmute_lens::<&mut Foo>();

        let mut query_a = lens.query();
        let mut query_b = lens.query();
        //~^ E0499

        let a = query_a.single_mut();
        let b = query_b.single_mut(); // oops 2 mutable references to same Foo
        assert_eq!(*a, *b);
    }
}
