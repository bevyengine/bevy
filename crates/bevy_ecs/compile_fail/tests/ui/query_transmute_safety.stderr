error[E0499]: cannot borrow `query` as mutable more than once at a time
  --> tests/ui/query_transmute_safety.rs:19:26
   |
18 |         let mut lens_a = query.transmute_lens::<&mut Foo>();
   |                          ----- first mutable borrow occurs here
19 |         let mut lens_b = query.transmute_lens::<&mut Foo>();
   |                          ^^^^^ second mutable borrow occurs here
...
22 |         let mut query_a = lens_a.query();
   |                           ------ first borrow later used here

error[E0499]: cannot borrow `lens` as mutable more than once at a time
  --> tests/ui/query_transmute_safety.rs:34:27
   |
33 |         let mut query_a = lens.query();
   |                           ---- first mutable borrow occurs here
34 |         let mut query_b = lens.query();
   |                           ^^^^ second mutable borrow occurs here
...
37 |         let a = query_a.single_mut();
   |                 ------- first borrow later used here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0499`.
